### DISCUSSION TITLE
Heterogeneous node types with homogeneous edges

### DISCUSSION THREAD

**Discussion by JihoChoi** [pyg-team/pytorch_geometric#5325] [Q&A] [answered]
What would be the correct way of building a custom graph which contains `heterogeneous node types` with `homogeneous edges`?

Currently, nodes of **type 1** have labels, and nodes of **type 2** don't have labels.
When I use train_mask (which has `True` for type 1 node indexes), the whole type 2 nodes are deleted. [1]
However, if I use `HeteroData` the `type2-type2` edges are two same which leads to not good results. [2]
Eventually, I would like to pass the loss on the type 1 nodes label and keep the massage passing to go through the type 2 nodes.

Could someone suggest a way of handling this? (maybe using `LinkNeighborLoader`)

Thank you.

[1]
```
train_loader = NeighborLoader(  # LinkNeighborLoader
    dataset['train'],
    num_neighbors=[20] * 1,
    batch_size=1024,
    input_nodes=torch.tensor(dataset['train'].train_mask),
    shuffle=True,
)
```

[2]
```
data['type_1', 'transfers', 'type_1'].edge_index = edge_index[dataset_name]['11']
data['type_1', 'transfers', 'type_2'].edge_index = edge_index[dataset_name]['12']
data['type_2', 'transfers', 'type_1'].edge_index = edge_index[dataset_name]['21']
data['type_2', 'transfers', 'type_2'].edge_index = edge_index[dataset_name]['22']
...

train_loader = NeighborLoader(
    dataset['train'],
    num_neighbors={key: [20] * 4 for key in dataset['train'].edge_types},for 2 iterations
    batch_size=1024,
    input_nodes=('type_1', None),
    shuffle=False,
)
```

**Comment by rusty1s** [comment, marked as answer]
[1] I don't think nodes of type 2 are deleted. They will get properly used when constructing subgraphs. Let me know if I am misunderstanding something
[2] You can model it that way as well which gives you more fine-grained control on how you want to treat different node types. Why does it lead to bad results?
